src/command_modules/azure-cli-iot/azure/cli/command_modules/iot/sas_token_auth.py

```diff
@@ -0,0 +1,57 @@
+#---------------------------------------------------------------------------------------------
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License. See License.txt in the project root for license information.
+#---------------------------------------------------------------------------------------------
+
+from base64 import b64encode, b64decode
+from hashlib import sha256
+from hmac import HMAC
+from time import time
+try:
+    from urllib import (urlencode, quote)
+except ImportError:
+    from urllib.parse import (urlencode, quote)  # pylint: disable=import-error,no-name-in-module
+from msrest.authentication import Authentication
+
+
+class SasTokenAuthentication(Authentication):
```

Comment:
Yes, you are right. The reason I put it here is to distinguish from the auto-generated code.
This authentication is for the IoT Hub Device registry rest API (swagger_iot_hub_device_identity.json), which is not really Azure resource management API. And it use Shared Access Signature to do authorization. But msrest package only has basic auth, basic token auth and OAuth. So we have to implement it on our own.
You can read more details about SAS at [here](https://azure.microsoft.com/en-us/documentation/articles/iot-hub-sas-tokens/).